source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,188 @@ unittest
+            auto slot_idx = this.ledger.getBlockHeight() + 1;
+            this.nominate(slot_idx, data);
+        }
+    }
+
+    static class MisbehavingValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new BadNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class OneTimeNominator : Nominator
+    {
+        private Ledger ledger;
+        private bool is_nominating;
+
+        /// Ctor
+        public this (NetworkManager network, KeyPair key_pair,
+            Ledger ledger, TaskManager taskman)
+        {
+            this.ledger = ledger;
+            super(network, key_pair, ledger, taskman);
+        }
+
+        ///
+        public override bool isNominating () @safe @nogc nothrow
+        {
+            return this.is_nominating;
+        }
+
+        ///
+        public override void tryNominate () @safe
+        {
+            if (this.ledger.getBlockHeight < 1)
+            {
+                super.tryNominate();
+            }
+        }
+    }
+
+    static class OneTimeValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new OneTimeNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        ///
+        public this (immutable(Block)[] blocks)
+        {
+            super(blocks);
+        }
+
+        /// see base class
+        public override void createNewNode (Config conf)
+        {
+            if (this.nodes.length == 0)
+            {
+                auto api = RemoteAPI!TestAPI.spawn!MisbehavingValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+            else
+            {
+                auto api = RemoteAPI!TestAPI.spawn!OneTimeValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+        }
+    }
+
+    auto network = makeTestNetwork!BadAPIManager(TestConf.init);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
```

Comment:
```suggestion
    network.start();
    scope(exit) network.shutdown();
    scope(failure) network.printLogs();
```

Because if the network hasn't started, calling `shutdown` / `printLogs` is invaldi